veps/sig-compute/10-dra-devices/vep.md

```diff
@@ -0,0 +1,760 @@
+```go
+// DeviceStatus has the information of all devices allocated spec.domain.devices
+type DeviceStatus struct {
+    // PodName is the name of the virt-launcher that these devices belong to
+	// In case of live-migration there could be more than one pods in which case
+	// dra controller will look at the target pod and update its information
+	// the source pod information will be lost. At any give time, there should only be 
+	// one pod reflected here and true holder of device resources
+	PodName *string `json:"gpuStatuses,omitempty"`
+	// GPUStatuses reflects the state of GPUs requested in spec.domain.devices.gpus
+	// +listType=atomic
+	// +optional
+	GPUStatuses []DeviceStatusInfo `json:"gpuStatuses,omitempty"`
+	// HostDeviceStatuses reflects the state of GPUs requested in spec.domain.devices.hostDevices
+	// +listType=atomic
+	// +optional
+	HostDeviceStatuses []DeviceStatusInfo `json:"hostDeviceStatuses,omitempty"`
+}
+
+type DeviceStatusInfo struct {
+	// Name of the device as specified in spec.domain.devices.gpus.name or spec.domain.devices.hostDevices.name
+	Name string `json:"name"`
+	// DeviceResourceClaimStatus reflects the DRA related information for the device
+	DeviceResourceClaimStatus *DeviceResourceClaimStatus `json:"deviceResourceClaimStatus,omitempty"`
+}
+```
+
+Steps to make live migration work:
+1. DRA Status controller will include looking at target pods updating the above fields with device information
+2. Migration Controller currently, moves the migration from Scheduling to Scheduled when the target pod is running. At
+   this point the dra controller should have update the target pod information. Migration Controller should assert that
+   DRA status it sees is from the target pod by comparing the target pod name with podName in the status. The changes
+   should go here: https://github.com/kubevirt/kubevirt/blob/ada65cb7d99033baa3c096820027d08532e25c1e/pkg/virt-controller/watch/migration/migration.go#L624
+3. Once the above is asserted, the migration will continue to flow in the same way as it does today. When target
+   virt-handler will call SyncVMI on the virt-launcher, the virt-launcher will use the same code to convert the
+   deviceStatus into domxml.
+
+
+## Scalability
+
+1. virt-controller will have an additional control loop, however this will use the same shared informers of virt-controller
+   resulting in 0 additional watch calls
+1. virt-controller watch events will be properly filtered such that only changes to relevant status sections of pod and
+   vmi will trigger reconcile loop. Additional metrics for the workqueue of this control loop will be exposed
+1. At max virt-controller should generate 2 additional patch calls to vmi status in a happy path
+
+## Update/Rollback Compatibility
+
+- Since the feature will be behind a feature gate updates will not be impacted
+- Since the feature will be behind a feature gate, rollback will continue to work as long as feature gate is disabled
```

Comment:
What about when the feature gate GA?

Reply:
In the current version of this document, update/rollback compatibility is only filled out for alpha stage of the feature. When this feature is in GA, this section will be updated accordingly. 

Is that correct implementation of the VEP process? atleast that is how I have done it with couple of Kubernetes KEPs

Reply:
I think we should at least have a plan for when the feature gate gets graduated out.
Otherwise this entire document could be "don't worry about this, it's all behind a feature gate anyway!" :)
If any part of this design just doesn't possibly play well with upgrades, it'd be great to find out now and not later (even if API changes of alpha features are fair game).
This is why we expressed concerns about that string being moved to an inline struct for example.

Reply:
updated as requested, PTAL